helix-term/src/application.rs

```diff
@@ -420,6 +420,13 @@ impl Application {
                         }
                     }
                     Notification::PublishDiagnostics(params) => {
+                        // Insert the original lsp::Diagnostics here because we may have no open document
+                        // for diagnosic message and so we can't calculate the exact position.
+                        // When using them later in the diagnostics picker, we calculate them on-demand.
+                        self.editor
+                            .diagnostics
+                            .insert(params.uri.clone(), params.diagnostics.clone());
+
```

Comment:
This adds a whole bunch of duplication. Can we just store diagnostics on `editor` instead of `document`? I guess one painpoint is that raw lsp diagnostics don't have the offsets translated which makes this hard for files that haven't been opened yet.

Reply:
You mean changing the local, already location-converted diagnostics to use the `self.editor.diagnostics` list and convert the locations on demand like it is done in the diags-picker?

Reply:
It's expensive to do that on the fly. How about using a map for diagnostics of buffers that aren't currently open? Then if the document gets opened you remove the entry from the map, convert the values and store them in the document.

Reply:
How about using the `editor.diagnostics` map as single source of truth storing both like this? So we can cache already converted diags in a central place. 

```rust
enum DiagnosticType {
    Hx(Diagnostic),
    Lsp(LspDiagnostic)
}

BTreeMap<lsp::Url, Vec<DiagnosticType>>
```

BTW: Another issue is the missing highlightling of the previews that have no open document. Are you planning to solve this? What would be a good way of solving this without huge performance impact?

Reply:
> How about using the `editor.diagnostics` map as single source of truth storing both like this? So we can cache already converted diags in a central place.

I think that's a smart way of doing it, with diagnostics being converted from `LanguageServer` to `Local` when possible.



> BTW: Another issue is the missing highlightling of the previews that have no open document. Are you planning to solve this? What would be a good way of solving this without huge performance impact?

Yeah there's an issue open, I think we'll want to trigger highlighting after an idle timeout. That way it doesn't highlight hundreds of files if you're scrolling, but it will highlight the file you stop on.